src/messaging/ExchangeContext.cpp

```diff
@@ -0,0 +1,390 @@
+
+#include <inttypes.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#include <core/CHIPCore.h>
+#include <core/CHIPEncoding.h>
+#include <messaging/ExchangeContext.h>
+#include <messaging/ExchangeMgr.h>
+#include <protocols/CHIPProtocols.h>
+#include <protocols/common/CommonProtocol.h>
+#include <support/logging/CHIPLogging.h>
+#include <system/SystemTimer.h>
+
+using namespace chip::Encoding;
+using namespace chip::Inet;
+using namespace chip::System;
+
+namespace chip {
+
+// Special node id values.
+static constexpr NodeId kNodeIdNotSpecified = 0ULL;
+static constexpr NodeId kAnyNodeId          = 0xFFFFFFFFFFFFFFFFULL;
+
+static void DefaultOnMessageReceived(ExchangeContext * ec, const PacketHeader & packetHeader, uint32_t protocolId, uint8_t msgType,
+                                     PacketBuffer * payload)
+{
+    ChipLogError(ExchangeManager, "Dropping unexpected message %08" PRIX32 ":%d %04" PRIX16 " MsgId:%08" PRIX32, protocolId,
+                 msgType, ec->GetExchangeId(), packetHeader.GetMessageId());
+
+    PacketBuffer::Free(payload);
+}
+
+bool ExchangeContext::IsInitiator() const
+{
+    return mFlags.Has(ExFlagValues::kFlagInitiator);
+}
+
+bool ExchangeContext::IsResponseExpected() const
+{
+    return mFlags.Has(ExFlagValues::kFlagResponseExpected);
+}
+
+void ExchangeContext::SetInitiator(bool inInitiator)
+{
+    mFlags.Set(ExFlagValues::kFlagInitiator, inInitiator);
+}
+
+void ExchangeContext::SetResponseExpected(bool inResponseExpected)
+{
+    mFlags.Set(ExFlagValues::kFlagResponseExpected, inResponseExpected);
+}
+
+CHIP_ERROR ExchangeContext::SendMessage(uint16_t protocolId, uint8_t msgType, PacketBuffer * msgBuf, uint16_t sendFlags,
+                                        void * msgCtxt)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    PayloadHeader payloadHeader;
+
+    // Don't let method get called on a freed object.
+    VerifyOrDie(mExchangeMgr != nullptr && mRefCount != 0);
+
+    // we hold the exchange context here in case the entity that
+    // originally generated it tries to close it as a result of
+    // an error arising below. at the end, we have to close it.
+    AddRef();
+
+    // Set the exchange ID for this header.
+    payloadHeader.SetExchangeID(mExchangeId);
+
+    // Set the protocol ID for this header.
+    payloadHeader.SetProtocolID(protocolId);
+
+    // Set the message type for this header.
+    payloadHeader.SetMessageType(msgType);
+
+    // If a response message is expected...
+    if ((sendFlags & kSendFlag_ExpectResponse) != 0)
+    {
+        // Only one 'response expected' message can be outstanding at a time.
+        VerifyOrExit(!IsResponseExpected(), err = CHIP_ERROR_INCORRECT_STATE);
+
+        SetResponseExpected(true);
+
+        // Arm the response timer if a timeout has been specified.
+        if (mResponseTimeout > 0)
+        {
+            err = StartResponseTimer();
+            SuccessOrExit(err);
+        }
+    }
+
+    payloadHeader.SetInitiator(IsInitiator());
```

Comment:
Do we need to set initiator flag in every message that the initiator sends? Or, only in the first message from the initiator. (just curious)

Reply:
We need to set to every message, this field is used in message dispatch